### ISSUE TITLE
Discover HOST info only needed for `settings.waitForTcpPorts` ON

### ISSUE THREAD

**Issue by andresvia** [avast/gradle-docker-compose-plugin#230]
@augi 

I'm new to this project, if my understanding is correct, the "host" discovery happening, in `getContainerHost` is only needed if `settings.waitForTcpPorts` is ON, otherwise is not needed, I have a complex network setup, and I'm getting "Gateway cannot be obtained", I have comprehensive actual docker healthchecks that work perfect for me. I rather avoid the attempts to discover the docker host if is actually not going to be required to determine healthyness.

I may be able to send a PR later when I have some time, want to confirm that the approach/idea makes sense.

Thanks.

https://github.com/avast/gradle-docker-compose-plugin/blob/9375f1723e44c24548a65e986ac3547fe6c68c92/src/main/groovy/com/avast/gradle/dockercompose/DockerExecutor.groovy#L98

**Comment by augi** [comment]
Hello, the `host` is actually used everywhere. Or to be more specific - the plugin itself uses the discovered `host` value just for waiting for TCP ports. But this value is exposed to anyone and it is typically used (with the exposed port) as the address to access a service (e.g. running Postgres).

Could you please share how the `network` part of `docker inspect` output looks like in your case? Maybe this plugin doesn't handles a valid situation.

As hotfix, if you don't want to end up with the `Gateway cannot be obtained` error, set `SERVICES_HOST` environment variable before running Gradle.

**Comment by andresvia** [comment]
@augi , I did:

```
docker inspect $(docker-compose --file compose.conf ps --quiet) --format '{"{{.Id}}":{{json .NetworkSettings}}}' | jq .
```

See attachments. [networksetting.zip](https://github.com/avast/gradle-docker-compose-plugin/files/4571491/networksetting.zip)

On a dev box (Mac OSX) and on the CI environment, in MacOS X, composeUp works in the CI environment fails.

the plugin config is:

```
dockerCompose {
    useComposeFiles = ['compose.conf']
    waitForTcpPorts = false
    projectName = 'gb-workflows'
}
```

Both files looks identical to me, it seems like the mac logic short-circuits the discovery of host.

Let me know if you need more info to debug.

I guess this piece of information may be useful too, in docker-compose file I'm using this network configuration for all containers except "localbind".

```
    network_mode: service:localbind
```

This is a network mode in compose that emulates what the `docker run --network container:foo` 
does.

I guess this compose file may help you reproduce.

```
cat docker-compose.yaml 
version: '2.1'
services:
  localbind:
    image: alpine
    entrypoint: /bin/sleep
    command: infinity
    ports:
      - 8081:8081 # server1
      - 8082:8082 # server1
  server1:
    image: python:2
    command: python -m SimpleHTTPServer 8081
    network_mode: service:localbind
  server2:
    image: python:2
    command: python -m SimpleHTTPServer 8082
    network_mode: service:localbind
```

The way I'm harnessing this compose file is the following (not quite just as an example).

```
docker exec -it $(docker-compose ps --quiet server1 ) curl -I http://127.0.0.1:8082/
HTTP/1.0 200 OK
Server: SimpleHTTP/0.6 Python/2.7.18
Date: Sun, 03 May 2020 20:12:03 GMT
Content-type: text/html; charset=UTF-8
Content-Length: 768
```

compose service server1, "sees" compose service server2, as if they are in the same host.

This way all services communicate with each other as if they are running in the same host, and this is also seamless for devs that expect everything running on localhost.

Thanks @augi

**Comment by andresvia** [comment]
This is a contrived example but I think makes my point.

Assume the following docker compose file:

```
version: '3.0'
services:
  localbind:
    image: alpine
    entrypoint: /bin/sleep
    command: infinity
    ports:
      - 127.0.0.1:8080:8080 # for me HTTP
      - 192.168.65.3:8080:8443 # for peers on the same network HTTPS
  server1:
    image: python:2
    # assume python module SimpleSecureHTTPServer is a thing
    command: "python -m SimpleHTTPServer 8080 & python -m SimpleSecureHTTPServer 8443 & wait"
    network_mode: service:localbind
```

Then you could do

`curl -I http://127.0.0.1:8080/`

HTTP for yourself and for peers on the same network HTTPS

`curl --insecure -I https://192.168.65.3:8080/`

The "Ports" "NetworkSettings" inspect section for "localbind" is:

```
            "Ports": {
                "8080/tcp": [
                    {
                        "HostIp": "127.0.0.1",
                        "HostPort": "8080"
                    }
                ],
                "8443/tcp": [
                    {
                        "HostIp": "192.168.65.3",
                        "HostPort": "8080"
                    }
                ]
            }
```

**Comment by augi** [comment]
Thank you for your examples!

Using the right address to access an exposed endpoint is actually very tricky as we must support a lot of setups. E.g. bare Linux containers on Linux, Docker on Mac, Linux Containers on Windows, Windows Containers, also _fake_ Docker-In-Docker (actually Docker-next-to-Docker) is a very common scenario.

The core logic is in the [getContainerHost](https://github.com/avast/gradle-docker-compose-plugin/blob/master/src/main/groovy/com/avast/gradle/dockercompose/DockerExecutor.groovy#L98) and [getPortsMapping](https://github.com/avast/gradle-docker-compose-plugin/blob/master/src/main/groovy/com/avast/gradle/dockercompose/DockerExecutor.groovy#L151) methods.

**Comment by andresvia** [comment]
These are some relevant logs with some internal information masked out, also I'm grep'ing the relevant lines only.

```
Will use  as host of sam-local
```

```
localbind uses an image, skipping
Building sam-local
Creating MASKED_localbind_1 ... 
Creating MASKED_localbind_1 ... done
Creating MASKED_sam-local_1           ... 
Creating MASKED_sam-local_1           ... done
Starting process 'command 'docker-compose''. Working directory: MASKED Command: docker-compose --no-ansi -f compose.conf -p MASKED ps -q localbind
Starting process 'command 'docker-compose''. Working directory: MASKED Command: docker-compose --no-ansi -f compose.conf -p MASKED ps -q sam-local
Container ID of service localbind is 2fdb4449f9849be61739225a4202dd0d6cf48cc46b56d8c1e85006202947533a
Will use 172.19.0.1 (network MASKED_default) as host of localbind
Will use ServiceHost{host='172.19.0.1', type=NetworkGateway} as host of service localbind
Exposed TCP port on service 'localbind:3001' will be available as 3001
Exposed TCP port on service 'localbind:4575' will be available as 4575
Exposed TCP port on service 'localbind:4576' will be available as 4576
Exposed TCP port on service 'localbind:4583' will be available as 4583
Exposed TCP port on service 'localbind:4584' will be available as 4584
Exposed TCP port on service 'localbind:8000' will be available as 8000
Exposed TCP port on service 'localbind:8080' will be available as 8080
Exposed TCP port on service 'localbind:8083' will be available as 8083
Container ID of service sam-local is 5341a5caced854748b6aab8f0d9449800c49ab0c5b5b78ca9aa056845f053913
Will use  as host of sam-local
Starting process 'command 'docker-compose''. Working directory: MASKED Command: docker-compose --no-ansi -f compose.conf -p MASKED stop --timeout 10 localbind sam-local ... MASKED ...
Stopping MASKED_sam-local_1           ... 
Stopping MASKED_localbind_1           ... 
Stopping MASKED_sam-local_1           ... done
Stopping MASKED_localbind_1           ... done
Removing MASKED_sam-local_1           ... 
Removing MASKED_localbind_1           ... 
Removing MASKED_localbind_1           ... done
Removing MASKED_sam-local_1           ... done
```

Right after the exception is thrown.

Remember from my attachment before network inspect for all container except "localbind" is like:

```
{
    "Bridge": "",
    "EndpointID": "",
    "Gateway": "",
    "GlobalIPv6Address": "",
    "GlobalIPv6PrefixLen": 0,
    "HairpinMode": false,
    "IPAddress": "",
    "IPPrefixLen": 0,
    "IPv6Gateway": "",
    "LinkLocalIPv6Address": "",
    "LinkLocalIPv6PrefixLen": 0,
    "MacAddress": "",
    "Networks": {},
    "Ports": {},
    "SandboxID": "",
    "SandboxKey": "",
    "SecondaryIPAddresses": null,
    "SecondaryIPv6Addresses": null
  }
```

Let me also add a section of the stack trace which may be helpful.

```
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.RuntimeException: Gateway cannot be obtained
	at com.avast.gradle.dockercompose.DockerExecutor.getContainerHost(DockerExecutor.groovy:137)
	at com.avast.gradle.dockercompose.DockerExecutor$getContainerHost.call(Unknown Source)
	at com.avast.gradle.dockercompose.tasks.ComposeUp.createContainerInfo(ComposeUp.groovy:157)
	at com.avast.gradle.dockercompose.tasks.ComposeUp$_loadServicesInfo_closure12$_closure22.doCall(ComposeUp.groovy:151)
	at com.avast.gradle.dockercompose.tasks.ComposeUp$_loadServicesInfo_closure12.doCall(ComposeUp.groovy:151)
	at com.avast.gradle.dockercompose.tasks.ComposeUp.loadServicesInfo(ComposeUp.groovy:151)
	at com.avast.gradle.dockercompose.tasks.ComposeUp.up(ComposeUp.groovy:79)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
```

It may also be helpful to know that I'm using the 0.10.10 version of `gradle-docker-compose-plugin`.

@augi let me know if need more info.

**Comment by andresvia** [comment]
I think It may be helpful to know this too, `--network none` is a valid docker run flag,

and `network_mode: "none"` is a valid parameter for docker compose services.

**Comment by andresvia** [comment]
@augi would make "`SERVICES_HOST`" a setting like all the others (eg. `waitForTcpPorts`) have more sense/be cleaner?

**Comment by augi** [comment]
I've implemented some of the improvements in `0.12.0`. Unfortunately, custom port mapping with IP address specified (`ip_address:host_port:container_port`) is still not supported as it would require a major change in the plugin API (and I don't want to make backward-incompatible changes now).
